server/src/com/thoughtworks/go/config/update/PatchEnvironmentCommand.java

```diff
@@ -74,15 +78,50 @@ public void update(CruiseConfig configForEdit) throws Exception {
         for (String pipelineName : pipelinesToRemove) {
             environmentConfig.removePipeline(new CaseInsensitiveString(pipelineName));
         }
+
+        for (EnvironmentVariableConfig variableConfig : envVarsToAdd) {
+            environmentConfig.addEnvironmentVariable(variableConfig);
+        }
+
+        for (String variableName : envVarsToRemove) {
+            environmentConfig.getVariables().removeIfExists(variableName);
+        }
     }
 
     @Override
     public boolean isValid(CruiseConfig preprocessedConfig) {
         boolean isValid = validateRemovePipelines(preprocessedConfig);
         isValid = isValid && validateRemoveAgents(preprocessedConfig);
+        isValid = isValid && validateRemoveEnvironmentVariables(preprocessedConfig);
         return isValid && super.isValid(preprocessedConfig);
     }
 
+    private boolean validateRemoveEnvironmentVariables(CruiseConfig preprocessedConfig) {
+        EnvironmentConfig preprocessedEnvironmentConfig = preprocessedConfig.getEnvironments().find(environmentConfig.name());
+        if(preprocessedEnvironmentConfig instanceof MergeEnvironmentConfig){
+            for (String variableName : envVarsToRemove) {
+                if(preprocessedEnvironmentConfig.containsEnvironmentVariableRemotely(variableName)){
+                    String origin = ((MergeEnvironmentConfig) preprocessedEnvironmentConfig).getOriginForEnvironmentVariable(variableName).displayName();
+                    String message = String.format("Environment variable with name '%s' cannot be removed from environment '%s' as the association has been defined remotely in [%s]",
+                            variableName, environmentConfig.name(), origin);
+                    result.badRequest(actionFailed.addParam(message));
```

Comment:
This is a validation error. Usually `badRequest` comes into picture when 'The request could not be understood by the server due to malformed syntax.' Use 422 instead. See http://stackoverflow.com/questions/16133923/400-vs-422-response-to-post-of-data for an explanation. Same for pipelines and agents validation too.